File: com.top_logic.client.diagramjs/src/main/java/com/top_logic/client/diagramjs/model/util/MathUtil.java
/*
 * SPDX-FileCopyrightText: 2019 (c) Business Operation Systems GmbH <info@top-logic.com>
 * 
 * SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-BOS-TopLogic-1.0
 */
package com.top_logic.client.diagramjs.model.util;

import com.google.gwt.core.client.JavaScriptObject;

/**
 * Mathematical utilities.
 *
 * @author <a href="mailto:sfo@top-logic.com">Sven Förster</a>
 */
public class MathUtil {

	/**
	 * Euclidian distance between two positions.
	 */
	public static double getEuclidianDistance(Position position1, Position position2) {
		double xDifference = position1.getX() - position2.getX();
		double yDifference = position1.getY() - position2.getY();

		return Math.sqrt(xDifference * xDifference + yDifference * yDifference);
	}

	/**
	 * Position of the point we get if we move position1 t along the line between position1
	 *         and position2.
	 */
	public static Position getRelativeClosePoint(Position position1, Position position2, double t) {
		Position relativeClosePoint = JavaScriptObject.createObject().cast();

		relativeClosePoint.setX((1 - t) * position1.getX() + t * position2.getX());
		relativeClosePoint.setY((1 - t) * position1.getY() + t * position2.getY());

		return relativeClosePoint;
	}

	/**
	 * @see #getRelativeClosePoint(Position, Position, double)
	 */
	public static Position getAbsoluteClosePoint(Position position1, Position position2, double absoluteDistance) {
		double euclidianDistance = getEuclidianDistance(position1, position2);

		return getRelativeClosePoint(position1, position2, absoluteDistance / euclidianDistance);
	}

}

File: com.top_logic.umljs/src-js/lib/lib/draw/MathUtil.js
export function getEuclidianDistance(point1, point2) {
  var xDifference = point1.x - point2.x;
  var yDifference = point1.y - point2.y;

  return Math.sqrt(xDifference*xDifference + yDifference*yDifference);
};

export function getRelativeClosePoint(point1, point2, t) {
  return {
    x: (1-t)*point1.x + t*point2.x,
    y: (1-t)*point1.y + t*point2.y
  }
};

export function getAbsoluteClosePoint(point1, point2, absoluteDistance) {
  var distance = getEuclidianDistance(point1, point2);

  return getRelativeClosePoint(point1, point2, absoluteDistance / distance);
};
